src/mfaktc.c

```diff
@@ -612,24 +624,25 @@ other return value
 
     if(use_kernel == AUTOSELECT_KERNEL || use_kernel == UNKNOWN_KERNEL)
     {
-      printf("ERROR: No suitable kernel found for bit_min=%d, bit_max=%d.\n",
+      logprintf(mystuff, "ERROR: No suitable kernel found for bit_min=%d, bit_max=%d.\n",
                  mystuff->bit_min, mystuff->bit_max_stage);
       return RET_ERROR;
     }
   }
 
   sprintf(mystuff->stats.kernelname, "%s_%d", kernel_info[use_kernel].kernelname, mystuff->vectorsize);
 
-  if(mystuff->mode != MODE_SELFTEST_SHORT && mystuff->verbosity >= 1)printf("Using GPU kernel \"%s\"\n", mystuff->stats.kernelname);
+  if(mystuff->mode != MODE_SELFTEST_SHORT && mystuff->verbosity >= 1)logprintf(mystuff, "Using GPU kernel \"%s\"\n", mystuff->stats.kernelname);
 
   if(mystuff->mode == MODE_NORMAL)
   {
-    if((mystuff->checkpoints > 0) && (checkpoint_read(mystuff->exponent, mystuff->bit_min, mystuff->bit_max_stage, &cur_class, &factorsfound, mystuff->verbosity) == 1))
+    if((mystuff->checkpoints > 0) && (checkpoint_read(mystuff->exponent, mystuff->bit_min, mystuff->bit_max_stage, &cur_class, &factorsfound, mystuff->factors_string, &(mystuff->stats.bit_level_time), mystuff->verbosity) == 1))
```

Comment:
This doesn't compile on Linux, as `mystuff->stats.bit_level_time` is `unsigned long`, but the expected argument is `unsigned long long`. One way to fix it is to add a temporary variable of the expected type and then copy the value to `mystuff->stats.bit_level_time`.

Reply:
I had perhaps a bit naively ported over changes to mfaktc to here, and this is one point where they diverged, but MSC 19.40.33811 didn't complain, so I assumed it was perhaps okay as there was no error or warning.

mystuff in mfakto uses cl_ulong, which I don't fully understand the purpose of, but I figured I should continue the pattern. a ulong number of milliseconds is 5e8 years, so we shouldn't need a ulong long for this variable. I'll attempt to just change the type to ulong throughout.

Reply:
Delving deeper it seems like cl_ulong is defined as:
`typedef unsigned __int64        cl_ulong;`
which is the minimum size allocated to a c `unsigned long long` iirc.

So I guess we don't actually need to change either of the types, we only need to translate between them more smoothly like you say.

I am a bit surprised your compiler is saying that it's an unsigned long though. As we define it as a cl_ulong in my_types.h:
`cl_ulong bit_level_time;            /* time (in ms) since the bitlevel was started */`

Reply:
Does this mean your compiler is only using a 4 byte unsigned int for `cl_ulong`? but mine is using an 8 byte unsigned int for it?

Reply:
Here's the exact error message on 64-bit Linux:

```
mfaktc.c: In function ‘tf’:
mfaktc.c:639:168: error: passing argument 7 of ‘checkpoint_read’ from incompatible pointer type [-Wincompatible-pointer-types]
  639 |     if((mystuff->checkpoints > 0) && (checkpoint_read(mystuff->exponent, mystuff->bit_min, mystuff->bit_max_stage, &cur_class, &factorsfound, mystuff->factors_string, &(mystuff->stats.bit_level_time), mystuff->verbosity) == 1))
      |                                                                                                                                                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                                                                                                                                                        |
      |                                                                                                                                                                        cl_ulong * {aka long unsigned int *}
In file included from mfaktc.c:40:
checkpoint.h:21:154: note: expected ‘long long unsigned int *’ but argument is of type ‘cl_ulong *’ {aka ‘long unsigned int *’}
   21 | int checkpoint_read(unsigned int exp, int bit_min, int bit_max, unsigned int* cur_class, int *num_factors, char* factors_string, unsigned long long int* bit_level_time, int verbosity);
      |                                                                                                                                  ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~
```

Just because pointers point to variables of the same size it doesn't make them compatible. It should be the same type.

The gcc documentation says that `-Wincompatible-pointer-types` is an error for C99 and newer. I'm using gcc 14.2.1 that defaults to C17.